### ISSUE TITLE
Cargo parses Cargo.toml files in subdirectories unrelated to the current project

### ISSUE THREAD

**Issue by daboross** [rust-lang/cargo#1423]
If an invalid `Cargo.toml` file is created in a sub-directory for a cargo project, `cargo build` will fail even though the invalid `Cargo.toml` file is unrelated to the project.

Cargo will error  `An unknown error occurred` if there is a `subdir/Cargo.toml/` directory, and will error with `failed to parse manifest at ...` if an empty `subdir/Cargo.toml` file exists.

The commands I used to test, and the shell output from those commands:

```
$ cargo new test-project
$ cd test-project
$ mkdir subdir
$ touch subdir/Cargo.toml
$ cargo build
failed to parse manifest at `/home/daboross/test-project/subdir/Cargo.toml`

Caused by:
No `package` or `project` section found.
$ rm subdir/Cargo.toml
$ mkdir subdir/Cargo.toml
$ cargo build
An unknown error occurred

To learn more, run the command again with --verbose.
$ cargo build --verbose
Is a directory (os error 21)
```

**Comment by alexcrichton** [comment]
This is somewhat intended behavior due to the fact that we crawl repos for `Cargo.toml` instead of relying on a top-level manifest to be present. The behavior leaks into the main repository, though, and we could probably make the manifest loading more on-deman.

**Comment by daboross** [comment]
Ah, ok. It doesn't seem like the current cargo builds projects in sub-directories either though?

```
$ cargo new test
$ cargo build
Could not find `Cargo.toml` in `/tmp/tmp.VCKC532Yu2/` or any parent directory
$ tree -L 2
.
└── test
    ├── Cargo.toml
    ├── .git
    ├── .gitignore
    └── src

3 directories, 2 files
$
```

This is a directory where the only sub-directory is a `test` folder, and a cargo project exists in that folder - but cargo will still output an error saying it can't find a project in the current directory or parent directory.

**Comment by alexcrichton** [comment]
Today when you say something like:

```toml
[dependencies]
foo = { git = "https://github.com/bar/foo" }
```

We will crawl the entire structure of that repository looking for manifests. This means that when you depend on crate from that repo, the author can reorganize the repo how they like in a backwards compatible fashion.

This "deeply look for manifests" logic leaks into path dependencies (aka the local crate you're building) but it doesn't necessarily need to.

Note that the deep logic only kicks in when we already have a known path to point at. In your snippet, for example, we don't deeply search the filesystem for a manifest because we don't have any concrete path to go off yet!

**Comment by daboross** [comment]
Ahhh, ok - that makes sense, thanks!

**Comment by lucab** [comment]
I've recently hit this issue using [quilt](http://savannah.nongnu.org/projects/quilt), which stores patch status under `.pc`. I am not sure if git does something similar in specific cases (rerere maybe?), but I think it could also happen for .git/.svn./etc. I think we could at least agree that hidden directories should be excluded from inspection, to avoid other surprises. Are patches to avoid recursing in dot directories welcome?

**Comment by alexcrichton** [comment]
I'd personally prefer to fix the probing of path sources to not require probing the entire directory, although I realize that'd be more difficult than just ignoring dot directories.

**Comment by lucab** [comment]
When looking at the code, I realized that my issue is probably not exactly as the original reported here, and that Cargo code already accounts for some special cases to be ignored (.git/ and target/ dirs, in particular).

I'm still not familiar enough with Cargo code to fix the original issue, so for the moment I only expanded the existing special casing to all dot directories.

**Comment by jauhien** [comment]
Is it possible to add a possibility to explicitly deny cargo to look in subdirs? E.g. by specifying this in Cargo.toml?

Use case: I have a project where I want to add some examples compilable by cargo. They live in subdirectories. I do not want cargo to compile example by default, only the main project.

**Comment by alexcrichton** [comment]
To be clear, Cargo doesn't just build everything it finds, the bad behavior is that it's currently parsing `Cargo.toml` files (which could be invalid) and doing too much I/O (walking all the filesystem trees). Other than that though, I would prefer to solve this issue than add extra options to Cargo in the meantime.

**Comment by jauhien** [comment]
> To be clear, Cargo doesn't just build everything it finds

I see, that's why I edited my comment. ) Solving this issue would better than additional option, yes.

**Comment by badboy** [comment]
Any work on this yet? Would be nice to stop this behaviour at least when a manifest is explicitely specified using `--manifest-path`.

**Comment by alexcrichton** [comment]
@badboy unfortunately no, no progress here

**Comment by ehuss** [comment]
This seems to have been fixed around 1.9.  There have also been recent efforts to make the crawling for git repos more resilient as well.
